src/restruct_poc/model/model.hpp

```diff
@@ -15,64 +16,138 @@ static Node InvalidNode(-1, {0., 0., 0., 1., 0., 0., 0.});
 /// relationships between components in a turbine. Nodes represent points in space with
 /// position, displacement, velocity, and acceleration. Constraints represent relationships
 /// between nodes that restrict their relative motion in some way.
-struct Model {
-    std::vector<Node> nodes;
-    std::vector<Constraint> constraints;
+class Model {
+public:
+    Model() = default;
 
-    /// Adds a node to the model and returns the node
-    Node AddNode(
+    Model(std::vector<Node> nodes, std::vector<BeamElement> beam_elements) {
+        for (auto& node : nodes) {
+            this->nodes_.push_back(std::make_shared<Node>(node));
+        }
+        for (auto& element : beam_elements) {
+            this->beam_elements_.push_back(std::make_shared<BeamElement>(element));
+        }
+    }
+
+    Model(
+        std::vector<std::shared_ptr<Node>> nodes,
+        std::vector<std::shared_ptr<BeamElement>> beam_elements
+    )
+        : nodes_(std::move(nodes)), beam_elements_(std::move(beam_elements)) {}
+
+    // Make the class non-copyable and non-assignable
+    Model(const Model&) = delete;
+    Model& operator=(const Model&) = delete;
+
+    /// Add a node to the model and return a shared pointer to the node
+    std::shared_ptr<Node> AddNode(
         const Array_7& position, const Array_7& displacement = Array_7{0., 0., 0., 1., 0., 0., 0.},
         const Array_6& velocity = Array_6{0., 0., 0., 0., 0., 0.},
         const Array_6& acceleration = Array_6{0., 0., 0., 0., 0., 0.}
     ) {
-        this->nodes.push_back(Node(nodes.size(), position, displacement, velocity, acceleration));
-        return this->nodes.back();
+        auto node =
+            std::make_shared<Node>(nodes_.size(), position, displacement, velocity, acceleration);
+        this->nodes_.push_back(std::move(node));
+        return this->nodes_.back();
     }
 
-    /// Adds a rigid constraint to the model and returns the constraint.
-    Constraint AddRigidConstraint(const Node& node1, const Node& node2) {
-        this->constraints.push_back(
-            Constraint(ConstraintType::kRigid, this->constraints.size(), node1, node2)
-        );
-        return this->constraints.back();
+    /// Return a node by ID - const/read-only version
+    std::shared_ptr<const Node> GetNode(int id) const { return this->nodes_[id]; }
+
+    /// Return a node by ID - non-const version
+    std::shared_ptr<Node> GetNode(int id) { return this->nodes_[id]; }
```

Comment:
Should these return a const Node& and Node& respectively?  We never call this function where we don't immediately dereference the shared pointer, so we might as well dereference them in place.